### PR TITLE
fix testdiscoverysession telemetry 

### LINKED ISSUE #5688: Telemetry for discovery needs to be separated for unittest and pytest
attribute DefaultExecutorUri is controlling which string gets written to discovery telemetry 
We need 2 separate classes registered with Test Explorer to get separate unittest and pytest stats.

The downside is we will be parsing the settings twice to get the testframework 
image

image

### PR DESCRIPTION
**Pull request by bschnurr**
testdiscovery session telemetry data relies on DefaultExecutorUri so we need 2 separate classes to get separate telemetry for unittest and pytest

Fix #5688

### CODE REVIEW

File: Python/Product/TestAdapter.Executor/PythonTestDiscoverer.cs [hunk 1]
```diff
@@ -25,14 +25,39 @@
...
+    /// Note even though we we specify  [DefaultExecutorUri(PythonConstants.PytestExecutorUriString)] we still get all .py source files
     /// from all testcontainers.  
     /// </summary>
     [FileExtension(".py")]
     [DefaultExecutorUri(PythonConstants.PytestExecutorUriString)]
+    public class PytestTestDiscoverer : PythonTestDiscoverer {
+        public PytestTestDiscoverer() : base(TestFrameworkType.Pytest) { 
+        }
+    }
+
```
Comment (bschnurr) [inline Python/Product/TestAdapter.Executor/PythonTestDiscoverer.cs]: New separate classes for unittest and pytest

File: Python/Product/TestAdapter.Executor/Config/RunSettingsUtil.cs [hunk 2]
```diff
@@ -8,12 +8,13 @@
 
 namespace Microsoft.PythonTools.TestAdapter.Config {
     public class RunSettingsUtil {
-        public static Dictionary<string, PythonProjectSettings> GetSourceToProjSettings(IRunSettings settings) {
+        public static Dictionary<string, PythonProjectSettings> GetSourceToProjSettings(IRunSettings settings, TestFrameworkType filterType) {
```
Comment (bschnurr) [inline Python/Product/TestAdapter.Executor/Config/RunSettingsUtil.cs]: now that we have 2 separate classes being triggered for discovery we can do less work by skipping over projects that don't have the test framework we are looking for

File: Python/Product/TestAdapter.Executor/PythonTestDiscoverer.cs [hunk 3]
```diff
@@ -25,14 +25,39 @@
 
 namespace Microsoft.PythonTools.TestAdapter {
 
+    /// <summary>
+    /// Note even though we we specify  [DefaultExecutorUri(PythonConstants.UnitTestExecutorUriString)] we still get all .py source files
```
Comment (huguesv) [inline Python/Product/TestAdapter.Executor/PythonTestDiscoverer.cs]: we we

File: Python/Product/TestAdapter.Executor/PythonTestDiscoverer.cs [hunk 4]
```diff
@@ -25,14 +25,39 @@
...
+    [FileExtension(".py")]
+    [DefaultExecutorUri(PythonConstants.UnitTestExecutorUriString)]
+    public class UnitTestDiscoverer : PythonTestDiscoverer {
+        public UnitTestDiscoverer() : base(TestFrameworkType.UnitTest) {
+        }
+    }
 
     /// <summary>
-    /// Note even though we we specify  [DefaultExecutorUri(PythonConstants.PytestExecutorUri)] we still get all .py source files
+    /// Note even though we we specify  [DefaultExecutorUri(PythonConstants.PytestExecutorUriString)] we still get all .py source files
```
Comment (huguesv) [inline Python/Product/TestAdapter.Executor/PythonTestDiscoverer.cs]: we we

File: Python/Product/TestAdapter.Executor/PythonTestDiscoverer.cs [hunk 5]
```diff
@@ -25,14 +25,39 @@
...
+    }
+    
     public class PythonTestDiscoverer : ITestDiscoverer {
+        private TestFrameworkType _frameworkType;
+
+        /// <summary>
+        /// Create a test framework specific Test Discoverer
+        /// </summary>
+        /// <param name="frameworkType"></param>
+        public PythonTestDiscoverer(TestFrameworkType frameworkType) {
```
Comment (huguesv) [inline Python/Product/TestAdapter.Executor/PythonTestDiscoverer.cs]: not meant to be directly instantiated anymore, so you could make this ctor protected and mark the class abstract.

File: Python/Tests/TestAdapterTests/TestDiscovererTests.cs [hunk 6]
```diff
@@ -688,7 +689,20 @@ public abstract partial class TestDiscovererTests {
...
+            switch (testEnv.TestFramework) {
+                case FrameworkPytest:
+                    discoverer = new PytestTestDiscoverer();
+                    break;
+
+                case FrameworkUnittest:
+                    discoverer = new UnitTestDiscoverer();
+                    break;
+                default:
+                    Assert.Fail();
```
Comment (huguesv) [inline Python/Tests/TestAdapterTests/TestDiscovererTests.cs]: should pass in a message to `Fail`, "Unknown test framework"

**Pull request merged**